LibOS/shim/src/sys/shim_semget.c

```diff
@@ -837,7 +847,10 @@ int submit_sysv_sem (struct shim_sem_handle * sem, struct sembuf * sops,
     while (!sem_ops->stat.completed &&
            !sem_ops->stat.failed) {
         if (!sem->owned) {
-            list_del_init(&sem_ops->progress);
+            /* Chia-Che 10/17/17: sem_ops may move from semaphore to semaphore
+               base on its current state */
+            next_ops = &sem->sems[sem_ops->ops[sem_ops->stat.current].sem_num].next_ops;
+            listp_del_init(sem_ops, next_ops, progress);
```

Comment:
@donporter please check this code

sem_ops may move from semaphore to semaphore. This change should take care of this semantic.

Reply:
A quick review looks ok in terms of use of the primitives, but I am less confident of the semaphore implementation details.